cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -188,6 +188,18 @@ def _measurement_key_names_(self) -> AbstractSet[str]:
             for key in circuit_keys
         }
 
+    def _control_key_names_(self) -> AbstractSet[str]:
+        circuit_keys = [
+            value.MeasurementKey.parse_serialized(key_str)
+            for key_str in protocols.control_key_names(self.circuit)
+            # We don't expose control keys just used internally.
```

Comment:
A few things here:
1. Why do we need to distinguish between internal and external keys?
2. This misbehaves if the subcircuit has controlled ops before measurements with the same key, since those control keys would be external.
3. Please add test coverage for the internal-keys case (if we're keeping it)

Reply:
The main thing was repetitions. I didn't want to expose a million control keys just because it was repeated a million times.

But I'm realizing now that there's a deeper problem, that we may want some controlled ops in a subcircuit to be controlled by a measurement from the outer circuit, and some may be controlled by a local measurement. So some would need their key to be ":1"..":n" concatenated, and others would not. But IDK if there's a a-priori way to determine which is the case. (In fact it plays into circuit validation too; something that depends on a non-existent measurement key is fine for a subcircuit but not for the main circuit). Probably there's a simple answer for it, but I haven't had much time to think about it yet.

Reply:
Gave it some more thought, and I feel like this is the correct approach. One thought is for diagrams: diagrams will be built on that function and we don't want to show a million double-lines leading to the subcircuit because it is repeated a million times. But more generally, I think `control_keys(op)` should symbolize all the controls that you need to provide to the op: "what cbits do I need to make sure I've measured prior to running this operation". I'd consider the keys a subcircuit uses internally as an implementation detail and returning those as potentially confusing.

However I also think this logic needs to go into `mapped_circuit`, such that repetitions causes internal control keys to get subscripted, whereas external control keys are not. I'll add this and several tests, as it seems like it could be easy to get wrong.

Also, just to sanity check, my understanding is that point 2 above isn't applicable because that would imply the same key is measured to twice: once in the primary circuit and once in the subcircuit, which isn't currently allowed (unless I'm misunderstanding). We may have to revisit that possibility later if we ever do the multiple-measurements-per-key ask, but I think it's better to hold off and design that when the situation arises and we have more context.

Reply:
Ick, the required update to `mapped_circuit` seems to be impossible without some fairly large and invasive changes. The issue is that it's possible for an op to be conditional upon both a subcircuit-local measurement and an external measurement. So we'd need a way to rekey the local control key with the iteration (because its corresponding local measurement key will also be rekeyed), but not rekey the external control key (because the corresponding external measurement key is not rekeyed). 

In theory that's not too hard; we have all the information we need at the point we need to do it. But the problem is that our infrastructure for rekeying is all done at the op level via `protocols.with_key_path` (see [`rekey_op`](https://github.com/quantumlib/Cirq/blob/eb3d84c4cbf8782d9c2c06d807182dd767770efe/cirq-core/cirq/circuits/circuit_operation.py#L244)).

The nearest thing I can think for a solution is changing `rekey_op` from using `with_key_path` to using `with_measurement_key_mapping`, and pass in each key individually. But currently that protocol only takes string keys. IDK what happens if I just stringify keys and map them this way, or if there's a plan to update that protocol to include MeasurementKey objects. Alternately, maybe add an optional key name to `with_key_path` would do roughly the same thing.

I think I need some more input from you and @smitsanghavi regarding the current and future state of measurement keys to figure out what approach would preferred be here.

Alternately we could say that only local control keys are supported initially, so that we can continue to make progress.

Reply:
I might be missing some nuances here but I think the problem here stems from the fact that measurement keys are created and "owned" by the `MeasurementGate`. Until now, they were also used only within that gate and they get recorded into the measurement results log during execution. So this worked out well and we had a nice encapsulation.

But now the measurement is referred to by another operation. And it's still not existing in the global scope so the classically controlled operation relies (will rely, to be precise) on "name string matching" to link these semi-dynamically.

But the key is still owned by and localized to its "producer" `MeasurementGate`. So any updates to the key (name remapping, path adding, etc) are delegated to the gate. The complication arises when we try to reproduce these updates accurately outside the gate, where we might not have enough context (e.g. how many layers of nested CircuitOperations are between my controlled operation and the `MeasurementGate` that produced it).

I think the best solution here is to accept that we are expanding the scope of these classical bits. 

I propose we create `CBits` (all classical registers, not just "controls") as concretely as `Qubits`. Each `CBit` will be a `MeasurementKey` in an outer scope (`Circuits` in general but it could be an internal `FrozenCircuit` inside a`CircuitOperation`).`cbits` protocol will return all the `CBits` that an operation/entity deals with. A `MeasurementGate` will return the `MeasurementKey` it records and a classically controlled operation will return its `control_keys`. 

For remapping a key or adding a path inside a CircuitOperation now, the parent CircuitOperation will directly call `protocols.with_measurement_key_mapping` or `protocols.with_key_path` on the keys it has **direct access to** in its scope. It will then delegate `protocols.with_measurement_key_mapping({old_mkey: new_mkey})` to its constituent operations. So the remapping logic happens only once and in a scope that has all the context. All other operations are then "informed".

Reply:
The prerequisite here, as you suggest, is a `with_measurement_key_mapping` for `MeasurementKey` objects. This is part of [Phase 3 of MeasurementKeys RFC](https://tinyurl.com/structured-measurement-keys#heading=h.rpxzb8ck6d37) along with the `measurement_keys` protocol returning the actual objects instead of strings (waiting for a release and deprecation cycle). I will try to get that in ASAP.

Reply:
Rather than a new `with_keys_at_scope` protocol, it is probably better to add a `bound_keys` param to the existing `with_key_path` protocol. That describes the intent better. "Lift this subcircuit up to the specified key path, with these keys already bound".

Reply:
As an aside, I think a lot of these measurement_key/with_path/etc functions would be better as regular methods than protocols. Protocols seem better suited to when you really just want to be able to pass in some arbitrary thing and get a result out of it. But these key functions have more structure, and can just be members. Protocols make it harder to trace what's going on, with everything jumping in and out of protocol. I don't see the benefit of using a protocol for these.

Reply:
Cross-posted at https://github.com/quantumlib/Cirq/issues/4594#issuecomment-956404485, but I think also we should get rid of the ability to apply `parent_path` independently of the `mapped_circuit` function, and also remove the `deep=False` option in `mapped_circuit`. I think we've got too many substeps of subcircuit unrolling exposed publicly, when it should be a single function that does the whole thing at once.

I think the best approach for this would be pass an optional `parent_path` into `mapped_circuit`, and remove the `map_deep=False` option. This would give `mapped_circuit` all the information it needs to perform the unrolling and forward it through to sub-subcircuits etc as a single step.

If there's a need for shallow mapping then this probably won't work, and we'll need to continue to support lazy `parent_path` application. But I can't think of a use case where shallow mapping is useful. The only place in the code that uses it is an artificial test case.

(Note that we probably shouldn't get too addicted to circuit unrolling anyway: if/when we bring on `repeat-until` subcircuits, *no* `mapped_circuit` implementation will work; it would result in an infinite loop).

Reply:
@95-martin-orion @smitsanghavi Okay I have a private branch that seems like a mostly reasonable approach to this, though it may require some more discussion. However, I think it's best if we keep changes small and atomic. I propose that I update *this* PR to contain only functionality needed for local control keys (so `cirq.control_keys(subcircuit)` always returns empty set). Then a subsequent PR will add the functionality to pass external control keys into subcircuits, which isn't necessarily a blocker for classical control anyway. And it will be independent of PRs regarding the actual implementation of classical control gates, so it won't block progress if it ends up continuing to be thorny. Also I'll update the current PR to have `control_keys` protocol return objects instead of names. Sound reasonable?